### PR TITLE
Add Run Only On Vpn condition (fixes #928)

### LINKED ISSUE #928: "Connected to VPN" as a run condition.
I am a road warrior. I only want to sync to my syncthing at home when I am connected to VPN.

### PR DESCRIPTION
**Pull request by androw**
Implements #928
- [ ] Need minSdkVersion 23

### CODE REVIEW

**Review by Catfriend1 — CHANGES_REQUESTED** [review]
Thanks for helping here.

Should this default to true? 
```mPreferences.getBoolean(Constants.PREF_RUN_ONLY_ON_VPN, true);```

File: app/src/main/java/com/nutomic/syncthingandroid/service/Constants.java [hunk 1]
```diff
@@ -69,6 +70,10 @@ public static String DYN_PREF_OBJECT_CUSTOM_SYNC_CONDITIONS(String objectPrefixA
         return objectPrefixAndId + "_" + "custom_sync_conditions";
     }
 
+    public static String DYN_PREF_OBJECT_SYNC_ONLY_ON_WIFI(String objectPrefixAndId) {
```
Comment (Catfriend1) [inline app/src/main/java/com/nutomic/syncthingandroid/service/Constants.java:73]: ON_VPN?

**Review by Catfriend1 — COMMENTED** [review]
Thanks for helping here.

Should this default to true? 
```mPreferences.getBoolean(Constants.PREF_RUN_ONLY_ON_VPN, true);```

File: app/src/main/java/com/nutomic/syncthingandroid/service/Constants.java (hunk 1 above)
Comment (androw) [reply app/src/main/java/com/nutomic/syncthingandroid/service/Constants.java]: Both fixed!

**Comment by androw** [comment]
Did some testing on my side. We shouldn't merge this before moving to minSdkVersion 23.

**Comment by Zulgrib** [comment]
How do you tell if it is a real vpn or a fake vpn to filter ads served by localhost? 

The WiFi ssid restriction is a good example of how I would expect things to work as a user, but I don't know if something similar is possible.

**Comment by androw** [comment]
It's hard to cover every usecases. I have check rapidly and didn't find a way to check for the VPN name in a human interpretable form.
Let me know if you find something but I don't have a lot of time to work on this.

**Comment by Zulgrib** [comment]
VPN apps are supposed to advertise a specific intent, I don't know if it can be used to find and list them and/or if it is possible to query if that specific vpn software is currently providing the tunnel interface.

**Comment by androw** [comment]
Looks like we can now merge this

**Comment by Catfriend1** [comment]
How important is this? minSdkVersion 23
I remember we are on minSdk 21.

**Comment by androw** [comment]
Ooops looks like I messed up between local and main branches. Sorry :/

**Comment by Catfriend1** [comment]
No problem

**Comment by androw** [comment]
@kahhong you can find a debug build here: https://static.androw.eu/syncthing-debug.apk.zip
Feel free to test!

**Comment by Catfriend1** [comment]
I'm all ears for the result :)

**Comment by androw** [comment]
Friendly bump!

**Comment by Catfriend1** [comment]
How is the community feedback on this?

**Comment by kahhong** [comment]
Works like a charm (Syncthing 1.22.2-preview.1). 

Syncthing sleeps when VPN is disconnected.

Syncthing starts when VPN is connected.

**Comment by Catfriend1** [comment]
:-)

**Comment by kahhong** [comment]
Works like a charm (Syncthing 1.22.2-preview.1). 

Syncthing sleeps when VPN is disconnected.

Syncthing starts when VPN is connected.

Edit: test

**Comment by androw** [comment]
I've asked myself during implementation.
It's possible but it's one more option to add.
Maybe the better would be to rewrite the condition system to allow for cumulative or exclusive condition.

**Comment by Catfriend1** [comment]
@androw Yep, that is also on my mind. Many good wishes for the condition "system" are on the way looking at the issue tracker. Your solution is good and I'd like to make it available to users, but I'm pretty unsure about the easyness for new users to understand the "negated" condition.

**Comment by kahhong** [comment]
I did a bit of brainstorming and would like to share my thoughts. I think it would be diminishing returns to implement a fully functional bitwise logic operations (AND/OR/XOR) as I don't see any more network-related conditions to add.

Before "Run on VPN" condition was added, the run conditions were sufficient on an allow/deny basis (checked/unchecked). As far as i can tell, there was no bitwise logic involved. However, "Run on VPN" condition added more complexity to the logic of conditions pertaining to the network-related conditions of the phone.

The reasonable permuatations of the 3 network-related conditions (VPN, Mobile Data and WiFi) are as follows:
[Reasonable Combinations]
- Wifi only
- Mobile Data only
- Wifi AND Mobile Data only
- VPN AND Wifi only
- VPN AND Mobile Data only
- VPN AND (Mobile Data OR Wifi)

Hence, I'd like to suggest we condense these conditions on the UI:
- Run only on VPN
- Run on Wi-Fi
- Run on mobile data

into a dropdown options mentioned in the list of Reasonable Combinations.

@Catfriend1 @androw do peer review my logic pls, thanks!

**Comment by androw** [comment]
@Catfriend1 The "Run only on VPN" rule seems easy to understand.
@kahhong The 'I want to use VPN on mobile and I'm fine with VPN on my home wifi' rule seems an acceptable one and need a new condition "system". If I understand correctly, you're proposing to explicit the already existing possibilities?

**Comment by kahhong** [comment]
@androw I also agree that the "Run only on VPN" option is easy to understand.


I'm not sure what you meant by "explicit the already existing possibilities." I'm thinking since VPN has added a layer of complexity therefore, exploring other possibilities to meet in the middle can help provide advanced functionalities and keep it easy for the users at the same time.

**Comment by Catfriend1** [comment]
> The reasonable permuatations of the 3 network-related conditions (VPN, Mobile Data and WiFi) are as follows: [Reasonable Combinations]
> 
> * Wifi only
> * Mobile Data only
> * Wifi AND Mobile Data only
> * VPN AND Wifi only
> * VPN AND Mobile Data only
> * VPN AND (Mobile Data OR Wifi)
> 
> Hence, I'd like to suggest we condense these conditions on the UI:
> 
> * Run only on VPN
> * Run on Wi-Fi
> * Run on mobile data

I think that would be the way to go here. The implementation is fine, but "run only" was the way we've worked a lot in the past to avoid it on the UI. Let's go "whitelist without blacklist" strategy.

**Review by Catfriend1 — COMMENTED** [review]
See last comment

**Pull request closed without merging**